lib/results.go

```diff
@@ -276,27 +278,62 @@ type jsonResult Result
 // NewJSONEncoder returns an Encoder that dumps the given *Results as a JSON
 // object.
 func NewJSONEncoder(w io.Writer) Encoder {
-	var jw jwriter.Writer
+	var enc jwriter.Writer
 	return func(r *Result) error {
-		(*jsonResult)(r).MarshalEasyJSON(&jw)
-		if jw.Error != nil {
-			return jw.Error
+		(*jsonResult)(r).MarshalEasyJSON(&enc)
+		if enc.Error != nil {
+			return enc.Error
 		}
-		jw.RawByte('\n')
-		_, err := jw.DumpTo(w)
+		enc.RawByte('\n')
+		_, err := enc.DumpTo(w)
 		return err
 	}
 }
 
 // NewJSONDecoder returns a Decoder that decodes JSON encoded Results.
-func NewJSONDecoder(r io.Reader) Decoder {
-	rd := bufio.NewReader(r)
+func NewJSONDecoder(rd io.Reader) Decoder {
+	dec := bufio.NewReader(rd)
 	return func(r *Result) (err error) {
 		var jl jlexer.Lexer
-		if jl.Data, err = rd.ReadBytes('\n'); err != nil {
+		if jl.Data, err = dec.ReadBytes('\n'); err != nil {
 			return err
 		}
 		(*jsonResult)(r).UnmarshalEasyJSON(&jl)
 		return jl.Error()
 	}
 }
+
+type GQLResponse struct {
+	Data   interface{}
+	Errors []GQLError
+}
+
+type GQLError struct {
+	Extensions interface{}
+	Message    string
+}
+
+// AsGraphQL re-interprets the given Result with GraphQL semantics, mutating it accordingly
+func AsGraphQL(r *Result) error {
+	if r.Code < 200 || r.Code >= 400 {
+		return nil
+	}
+
+	var res GQLResponse
+	err := json.Unmarshal(r.Body, &res)
+	if err != nil {
+		return err
+	}
+
+	if res.Errors != nil && len(res.Errors) > 0 {
+		for i, e := range res.Errors {
+			if i == 0 {
+				r.Error = e.Message
+			} else {
+				r.Error = fmt.Sprintf("%v, %v", r.Error, e.Message)
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
@tsenart, let me know if this is more along the lines of what you were hoping to see. I left `AsGraphQL()` here so I could keep the test checking for mutations of the result object.